kernel/src/log_compaction/writer.rs

```diff
@@ -0,0 +1,265 @@
+            self.min_file_retention_timestamp_millis,
+            None, // txn_expiration_timestamp
+        );
+
+        // Process actions using the same iterator pattern as checkpoints
+        // The processor handles reverse chronological processing internally
+        let result_iter = processor.process_actions_iter(actions_iter);
+
+        // Collect processed batches into a vector for the iterator
+        let mut batches = Vec::new();
+        let mut total_actions = 0i64;
+        let mut total_add_actions = 0i64;
+
+        for batch_result in result_iter {
+            let batch = batch_result?;
+            // Track statistics
+            total_actions += batch.actions_count;
+            total_add_actions += batch.add_actions_count;
+            // Extract the data from the batch - pass the FilteredEngineData directly
+            batches.push(batch.filtered_data.data);
+        }
+
+        Ok(LogCompactionDataIterator {
+            batches: batches.into_iter(),
+            total_actions,
+            total_add_actions,
+        })
+    }
+
+    /// Finalize the compaction after the data has been written
+    ///
+    /// This method completes the compaction process and can perform any
+    /// necessary cleanup or validation.
+    #[allow(dead_code)]
+    pub(crate) fn finalize(
+        self,
+        _engine: &dyn Engine,
+        _metadata: &FileMeta,
+        _data_iterator: LogCompactionDataIterator,
+    ) -> DeltaResult<()> {
+        // For now, just validate that the iterator was exhausted
+        // In the future, we might want to validate file metadata, etc.
+        Ok(())
+    }
+
+    /// List commit files in the specified version range
+    fn list_commit_files(&self, engine: &dyn Engine) -> DeltaResult<Vec<ParsedLogPath<FileMeta>>> {
+        let log_root = self.table_root.join("_delta_log/")?;
+
+        // Get all files in the _delta_log directory
+        let log_files = engine.storage_handler().list_from(&log_root)?;
+
+        // Filter and parse log files, keeping only commits in our version range
+        let mut commit_files = Vec::new();
+
+        for file in log_files {
+            let file = file?; // Handle the Result<FileMeta, Error>
+            if let Some(parsed_path) = ParsedLogPath::try_from(file)? {
+                if parsed_path.is_commit()
+                    && parsed_path.version >= self.start_version
+                    && parsed_path.version <= self.end_version
+                {
+                    commit_files.push(parsed_path);
+                }
+            }
+        }
+
+        // Sort by version in ascending order (LogSegment expects this)
+        // Note: Reverse chronological processing should happen in the log replay processor
+        // to ensure latest actions win during reconciliation
+        commit_files.sort_by_key(|f| f.version);
+
+        Ok(commit_files)
+    }
+
+    /// Create a log segment from the commit files
+    pub(crate) fn create_log_segment(
```

Comment:
can we use the existing `LogSegment` constructors (perhaps `for_table_changes` which will hopefully soon be generalized to `build_incremental`)

see #1159 for some related work on that front

Reply:
I think it is better to just inject `snapshot` into the writer and use the `LogSegment` that is already calculated. I need to do some refactoring to enable this. Raised https://github.com/delta-io/delta-kernel-rs/pull/1264 and converting this PR into draft until  https://github.com/delta-io/delta-kernel-rs/pull/1264 is merged.